### ISSUE TITLE
Why is ondelete_callback not called in Models?

### ISSUE THREAD

**Issue by Defcon0** [contao/core-bundle#1566]
Hello,

title says it all ;-) Shouldn't be the callback be fired? To my knowledge this only happens when deleting a record using a DC_Table action.

Thanks for clarification.

Bye

**Comment by aschempp** [comment]
Callback are only called in the DC (DC_Table), models do not have callbacks.

**Comment by Defcon0** [comment]
@aschempp I know that ;-) But what‘s the reason for that design decision?

**Comment by aschempp** [comment]
> But what‘s the reason for that design decision?

I don't think there was an explicit design decision. Callbacks existed back in Contao 2, models were added in Contao 3. Most of the callback arguments to not make sense in models, e.g. there is no `DataContainer` object to be passed.

**Comment by Defcon0** [comment]
OK, but I personally think, the callbacks should be triggered, since the models in fact are built on top of the database entities built using the `*Data Container* Array (DCA)`. In fact I assumed that this is the case until I found out last week that it's not. Building a data container out of the model data and passing it into the callback invocation in the `Model::delete()` function (and the other ones like `save()`) should be possible, or is it more complicated than it is on the first sight?

**Comment by Defcon0** [comment]
I agree with you, it would be better to migrate to Doctrine Entities if Contao now makes more and more use of Symfony. I understand your concerns, but at the moment I think it's an inconsistency in the framework that can have unwanted side effects like the following situation:

The developer has defined some trigger email for the customer in `ondelete_callback` and relies on the presence of the invocation of the DCA callbacks when using models operating on the DCA entities where the callbacks are defined. In the end this callbacks aren't triggered and the important code hasn't been called unintendedly.

Having a clear and consistent workflow with Doctrine Enties based on the DCA would be awesome, but you guys have more insight in contao's versioning constraints and the roapmap, of course. Maybe it can be planned in for Contao 5 (here BC breaks anyway). Maybe writing an automatic migration from model to doctrine entities _can_ be an option if callback invocation is deactivated as a default (which won't be the case in Contao 5 anymore). But probably packing it into Contao 5 would cause less dev pain I suspect ;-)
